syntaxes/haskell.YAML-tmLanguage

```diff
@@ -573,6 +547,35 @@ repository:
           | }             # A block closed? Maybe this should also include `;`, because non-indentation based `do`
         )
         | ^(?!\1\s+\S|\s*$)    # at least one, no-further indented, non-whitespace character. I.e. a same-level declaration/implementation
+  type_application:
+    patterns:
+      - begin: '(?<=\s)(@)(\()'
+        beginCaptures: 
+          '1': {name: keyword.operator.at.haskell}
+          '2': {name: punctuation.paren.haskell}
+        end: '\)'
+        endCaptures:
+          '0': {name: punctuation.paren.haskell}
+        name: meta.type-application.haskell
+        patterns:
+          - include: '#type_signature'
+      - begin: '(?<=\s)(@)\s*(\[)'
```

Comment:
now that I think about it ... do you think we could somehow only trigger on the `@` (plus look behind) and then dispatch into another set of `begin`-`end` rules for `[ ]`, `( )` and `ident` expressions? The termination condition for the outer multiline match would be the concatenation of the inner end conditions, perhaps as a look-behind. This way we don't have to duplicate the complicated start condition so many times? I think it could actually work.

Reply:
I haven't been able to figure this one out. Feel free to give it a go, otherwise I think I'll just use the current solution.